core/drivers/clk/clk.c

```diff
@@ -0,0 +1,615 @@
+		return clk_scmi_dummy_rates(clk, start_index, rates, nb_elts);
+
+	if (clk->ops->get_rates_array(clk, start_index, rates, nb_elts))
+		return SCMI_GENERIC_ERROR;
+
+	return SCMI_SUCCESS;
+}
+#endif
+
+#if defined(CFG_DT)
+
+struct clk *clk_of_get_simple_clk(struct clk_of_phandle_args *args __unused,
+				  void *data)
+{
+	return data;
+}
+
+struct clk_of_provider {
+	int nodeoffset;
+	struct clk *(*get_of_clk)(struct clk_of_phandle_args *args, void *data);
+	void *data;
+	SLIST_ENTRY(clk_of_provider) link;
+};
+
+static SLIST_HEAD(, clk_of_provider) clk_of_provider_list =
+				SLIST_HEAD_INITIALIZER(clk_of_provider_list);
+
+int clk_of_register_clk_provider(int nodeoffset,
+				 struct clk *(get_of_clk)(struct clk_of_phandle_args *, void *),
+				 void *data)
+{
+	struct clk_of_provider *prv;
+
+	prv = malloc(sizeof(*prv));
+	if (!prv)
+		return -1;
+
+	prv->get_of_clk = get_of_clk;
+	prv->data = data;
+	prv->nodeoffset = nodeoffset;
+
+	SLIST_INSERT_HEAD(&clk_of_provider_list, prv, link);
+
+	return 0;
+}
+
+static struct clk_of_provider *clk_get_provider(int nodeoffset)
+{
+	struct clk_of_provider *prv;
+
+	SLIST_FOREACH(prv, &clk_of_provider_list, link) {
+		if (prv->nodeoffset == nodeoffset)
+			return prv;
+	}
+
+	return NULL;
+}
+
+static uint32_t fdt_clock_cells(void *fdt, int nodeoffset)
+{
+	const fdt32_t *c;
+	int len;
+
+	c = fdt_getprop(fdt, nodeoffset, "#clock-cells", &len);
+	if (!c)
+		return len;
+
+	if (len != sizeof(*c))
+		return -FDT_ERR_BADNCELLS;
+
+	return fdt32_to_cpu(*c);
+}
+
+int clk_of_get_idx_by_name(void *fdt, int nodeoffset, const char *name)
+{
+	int idx;
+
+	idx = fdt_stringlist_search(fdt, nodeoffset, "clock-names", name);
+	if (idx < 0)
+		return -1;
+
+	return idx;
+}
+
+static struct clk *clk_of_get_from_provider(int nodeoffset,
+					    unsigned int clock_cells,
+					    const uint32_t *prop)
+{
+	unsigned int arg;
+	struct clk_of_phandle_args pargs;
+	struct clk_of_provider *prv;
+
+	pargs.args_count = clock_cells;
+	for (arg = 0; arg < clock_cells; arg++)
+		pargs.args[arg] = fdt32_to_cpu(prop[arg + 1]);
+
+	prv = clk_get_provider(nodeoffset);
+	if (!prv)
+		return NULL;
+
+	return prv->get_of_clk(&pargs, prv->data);
+}
+
+struct clk *clk_of_get_by_name(void *fdt, int nodeoffset,
+			       const char *name)
+{
+	int clk_id;
+
+	clk_id = clk_of_get_idx_by_name(fdt, nodeoffset, name);
+	if (clk_id < 0)
+		return NULL;
+
+	return clk_of_get_by_idx(fdt, nodeoffset, clk_id);
+}
+
+struct clk *clk_of_get_by_idx(void *fdt, int nodeoffset, int clk_idx)
+{
+	uint32_t phandle;
+	const uint32_t *prop;
+	int len, idx = 0, parent_node, clock_cells, idx32;
+
+	prop = fdt_getprop(fdt, nodeoffset, "clocks", &len);
+	if (!prop)
+		return NULL;
+
+	while (idx < len) {
+		idx32 = idx / sizeof(uint32_t);
+		phandle = fdt32_to_cpu(prop[idx32]);
+
+		parent_node = fdt_node_offset_by_phandle(fdt, phandle);
+		if (parent_node < 0)
+			return NULL;
+
+		clock_cells = fdt_clock_cells(fdt, parent_node);
+		if (clk_idx == 0)
```

Comment:
i don't really understand the sequence here.

Reply:
Since you can have clocks consuemr using multiple provider, they can have a different number of cells. For instance, you can have the following:

clocks = < &fixed_clk &pmc PMC_TYPE_PERIPHERAL 42 &pmc PMC_TYPE_PERIPHERAL 32>

In order to correctly gather the 3 clock for instance, you have to parse the clock property by progressing of the number of cells defined in the clock provider with #clock-cells. This loop does that, first compute that fixed_clk has 0 cells, add only the phandle size, then compute that the pmc has 2 #clock-cells, etc.